src/index.js

```diff
@@ -0,0 +1,224 @@
+import { exec } from 'child_process';
+import { sync as findUp } from 'find-up';
+import * as fs from 'fs-extra';
+import globCb from 'glob';
+import * as path from 'path';
+import semver from 'semver';
+import * as zkSnark from 'snarkjs';
+import { promisify } from 'util';
+
+const glob = promisify(globCb);
+
+zkSnark.bigInt.prototype.toJSON = function() {
+  return this.toString();
+};
+
+const Extension = {
+  Json: '.json',
+  Circom: '.circom',
+  VkProof: '.vk_proof',
+  VkVerifier: '.vk_verifier',
+  Solidity: '.sol'
+};
+
+class Snarks {
+  constructor(embark) {
+    this.embark = embark;
+    this.circuitsConfig = embark.pluginConfig;
+    this.compiledCircuitsPath = embark.config.dappPath('.embark', 'snarks');
+    // v5 detection won't work as desired with "dev embark" (built in monorepo)
+    // until after first prerelease of embark v5; can assign `true` to test
+    // this.isEmbark5 = true;
+    this.isEmbark5 = semver(this.embark.version).major >= 5;
+    this.registerEvents();
+  }
+
+  static circomBinary = findUp('node_modules/.bin/circom', { cwd: __dirname });
+  static snarkjsBinary = findUp('node_modules/.bin/snarkjs', {
+    cwd: __dirname
+  });
+
+  registerEvents() {
+    if (this.isEmbark5) {
+      return this.embark.registerActionForEvent(
+        'compiler:contracts:compile:before',
+        (contractFiles, callback) => {
+          this.compileAndGenerateContracts(contractFiles, callback);
```

Comment:
This is probably by design, but one thing that threw me off that in this code path we're passing `contractsFiles` while in the other we're passing an empty list. 

This sounds like two different outcomes depending on Embark version?

Reply:
Yes, for embark `<5.0.0` a different mechanism is used that doesn't involve adding to a list received by the callback, see line 203. It's a bit messy but I wasn't sure of a better way. I thought about creating methods that end in either `5` or `4`, i.e. so version-specific differences are contained in separate functions rather than handling it via conditionals in several functions, but I didn't convince myself that was a definitely better approach.

Reply:
That's fine, I'd also vote for keeping version numbers out of the code base and rather have them addressed by "feature". Sometimes we can't foresee if and how long something stays in a certain version(range), so it's better to keep them out of the equation from the start.